### DISCUSSION TITLE
Output logs from pyscipopt

### DISCUSSION THREAD

**Discussion by sheepone12138** [scipopt/PySCIPOpt#881] [Q&A] [answered]
I'm working on a project where I need to output the error during the iteration when the solver is running and plot it as a curve. How should I do this and how can I output the running log as a file?

**Comment by Joao-Dionisio** [comment]
Hello, @sheepone12138! When you say error, do you mean an error in SCIP, or some sort of error in your project?

If you know what is causing your error and when it's happening (in relation to the solving process), you can try implementing an event handler. For example, suppose your error shows up when solving the LP in some node. Then, in the event handler you need to catch the event SCIP_EVENTTYPE_LPSOLVED in ```eventinit```. You would then need to implement ```eventexec``` and plot the error you want or whatever you want to do.

[Here](https://www.scipopt.org/doc-6.0.0/html/type__event_8h.php) is a list of all events available in SCIP.

You can look [here](https://github.com/scipopt/PySCIPOpt/blob/f86fd95dae0f5def3d1874a52e3c57a8602e9c89/tests/test_tree.py#L5) for an example on using event handlers.


If you're having problem with this, you can tell me more details and I can try to help out.

To ouput the log into a file, you can try using [redirectOutput()](https://scipopt.github.io/PySCIPOpt/docs/html/classpyscipopt_1_1scip_1_1Model.html#a468d9a99d8f2cdef3dae0007f923fbbf). I've never used this, so I don't know how it works, sorry :(

**Reply by sheepone12138** [reply]
What I'm looking for is the curve where the error of SCIP gradually reduces during the iterative process of finding the correct answer, or output the following running log into a file format.
![50eb679668e94733bf11a1b34385ed05](https://github.com/user-attachments/assets/0cee72da-49c1-481a-ab08-4a5be8e0d29a)

**Reply by Joao-Dionisio** [reply]
Oh, you meant the gap, I understand now. Then yeah, for this you need an event handler that catches two events: ```SCIP_EVENTTYPE_BESTSOLFOUND``` and ``` SCIP_EVENTTYPE_LPSOLVED```. This can be a bit intimidating if you're not used to SCIP, but it's really not a big deal. Just an idea of what you need to do (I'm assuming it's a minimization problem):

```
from pyscipopt import Model, Eventhdlr, SCIP_EVENTTYPE

class gapEventhdlr(Eventhdlr):

    def eventinit(self): # we want to collect best primal solutions and best dual solutions
        self.model.catchEvent(SCIP_EVENTTYPE.BESTSOLFOUND, self)
        self.model.catchEvent(SCIP_EVENTTYPE.LPSOLVED, self)

    def eventexec(self, event):
        # if a new best primal solution was found, we save when it was found and also its objective
        if event == SCIP_EVENTTYPE.BESTSOLFOUND:
            self.model.data["primal_solutions"].append((self.model.getSolvingTime(), self.model.getObjVal()))
        elif event == SCIP_EVENTTYPE.LPSOLVED: # same for dual bound, but we need to check if it's better than the previous one
            if self.model.getDualbound() < self.model.data["dual_solutions"][-1][1]: # assuming minimization problem
                self.model.data["dual_solutions"].append((self.model.getSolvingTime(), self.model.getDualbound()))

m = Model()
m.data = {}
m.data["primal_solutions"] = []
m.data["dual_solutions"] = []
m.includeEventhdlr(gapEventhdlr, "gapEventHandler", "Event handler which collects primal and dual solution evolution")

# and then you build your model and just optimize. After optimizing, you can see what you want in m.data

```

I haven't tested this, so it might have some typos and the like, but it should be roughly correct. Let me know if you can follow it!

@mmghannam @Opt-Mucca, is this something you think should be added to the recipes? It does seem pretty annoying for users to have to implement for such a simple thing. Also, is there a reason for not having an event for BESTDUALSOLFOUND?

**Reply by sheepone12138** [reply]
Okay, thank you. I'll give your approach a try and let you know when I succeed

**Reply by mmghannam** [reply]
@Joao-Dionisio Sure yeah, but more broadly I think it can also be implemented more generally as an `attach_event_handler` method that takes some event types and a callback (function) to run when the event is dispatched (with the type and data passed to it as inputs). Much more ergonomic than writing a full class, which we can definitely handle internally.

**Reply by Joao-Dionisio** [reply]
Oh, super great idea!

**Reply by sheepone12138** [reply]
I tried your method and because of some bugs I changed your code to:
> gap_handler = GapEventhdlr()
>m.includeEventhdlr(gap_handler, "gapEventHandler", "Event handler which collects primal and dual solution evolution")

But then I ran into some mistakes that I couldn't figure out
>Traceback (most recent call last):
  File "C:\Users\15572\PycharmProjects\pythonProject2(scip)\sa.py", line 6, in eventinit
    self.model.catchEvent(SCIP_EVENTTYPE.BESTSOLFOUND)
  File "src\\pyscipopt\\scip.pxi", line 5014, in pyscipopt.scip.Model.catchEvent
TypeError: catchEvent() takes exactly 2 positional arguments (1 given)
Exception ignored in: 'pyscipopt.scip.PyEventInit'
Traceback (most recent call last):
  File "C:\Users\15572\PycharmProjects\pythonProject2(scip)\sa.py", line 6, in eventinit
    self.model.catchEvent(SCIP_EVENTTYPE.BESTSOLFOUND)
  File "src\\pyscipopt\\scip.pxi", line 5014, in pyscipopt.scip.Model.catchEvent
TypeError: catchEvent() takes exactly 2 positional arguments (1 given)
[event.c:205] ERROR: Error <0> in function call
[set.c:5329] ERROR: Error <0> in function call
[scip_solve.c:407] ERROR: Error <0> in function call
[event.c:226] ERROR: event handler <gapEventHandler> not initialized
[set.c:5450] ERROR: Error <-8> in function call
[scip_solve.c:1853] ERROR: Error <-8> in function call
[scip_solve.c:3393] ERROR: Error <-8> in function call
[scip_solve.c:2358] ERROR: Error <-8> in function call
[scip_solve.c:2607] ERROR: Error <-8> in function call
Traceback (most recent call last):
  File "C:\Users\15572\PycharmProjects\pythonProject2(scip)\sa.py", line 35, in <module>
    m.optimize()
  File "src\\pyscipopt\\scip.pxi", line 3529, in pyscipopt.scip.Model.optimize
  File "src\\pyscipopt\\scip.pxi", line 280, in pyscipopt.scip.PY_SCIP_CALL
Exception: SCIP: method cannot be called at this time in solution process!

**Reply by Joao-Dionisio** [reply]
Ah, I forgot that you need self as well. So for catching the events you need to do ```self.model.catchEvent(SCIP_EVENTTYPE.BESTSOLFOUND, self)``` and ```self.model.catchEvent(SCIP_EVENTTYPE.LPSOLVED, self)```

I'll try to implement this myself and I'll get back to you when I get it working

**Reply by sheepone12138** [reply]
When I follow your method there is no more error, but what I don't quite understand is: when I use `print(m.data)`，The `primal_solutions` and `dual_solutions` obtained are empty sets, like this`{'primal_solutions': [], 'dual_solutions': []}`

**Comment by Joao-Dionisio** [comment, marked as answer]
Hey @sheepone12138, sorry for the absurdly large delay. I think this does what you want

```
from pyscipopt import Model, Eventhdlr, SCIP_EVENTTYPE, Eventhdlr

    class gapEventhdlr(Eventhdlr):

        def eventinit(self): # we want to collect best primal solutions and best dual solutions
            self.model.catchEvent(SCIP_EVENTTYPE.BESTSOLFOUND, self)
            self.model.catchEvent(SCIP_EVENTTYPE.LPSOLVED, self)

        def eventexec(self, event):
            # if a new best primal solution was found, we save when it was found and also its objective
            if event.getType() == SCIP_EVENTTYPE.BESTSOLFOUND:
                self.model.data["primal_solutions"].append((self.model.getSolvingTime(), self.model.getPrimalbound()))
            
            if not self.model.data["dual_solutions"]:
                self.model.data["dual_solutions"].append((self.model.getSolvingTime(), self.model.getDualbound()))
            
            if self.model.getObjectiveSense() == "minimize":
                if self.model.isGT(self.model.getDualbound(), self.model.data["dual_solutions"][-1][1]):
                     self.model.data["dual_solutions"].append((self.model.getSolvingTime(), self.model.getDualbound()))
            else:
                if self.model.isLT(self.model.getDualbound(), self.model.data["dual_solutions"][-1][1]):
                    self.model.data["dual_solutions"].append((self.model.getSolvingTime(), self.model.getDualbound()))


    m = YOUR MODEL HERE
    m.data = {}
    m.data["primal_solutions"] = []
    m.data["dual_solutions"] = []

    hdlr = gapEventhdlr()
    m.includeEventhdlr(hdlr, "gapEventHandler", "Event handler which collects primal and dual solution evolution")

    m.optimize()
    m.data["primal_solutions"].append((m.getSolvingTime(), m.getPrimalbound()))
    m.data["dual_solutions"].append((m.getSolvingTime(), m.getDualbound()))
    print(m.data)
```

Using the model from helpers with a 5-second limit:

```
    from helpers.utils import bin_packing_model
    from random import randint

    m = bin_packing_model(sizes=[randint(1,40) for _ in range(120)],  capacity=50)
    m.setParam("limits/time",5)
```

And this matplotlib code

```
    from matplotlib import pyplot as plt

    time_primal, val_primal = zip(*m.data["primal_solutions"])
    plt.plot(time_primal, val_primal, label="Primal bound")
    time_dual, val_dual = zip(*m.data["dual_solutions"])
    plt.plot(time_dual, val_dual, label="Dual bound")

    plt.legend(loc="best")
    plt.show()
```


I get this

![image](https://github.com/user-attachments/assets/e39b1f3e-b5e0-4f84-b16a-27088593907d)

It's what you wanted, right?

**Reply by sheepone12138** [reply]
I will try to use it. Thanks again for your great help.

**Reply by sheepone12138** [reply]
I am truly sorry. When I was trying to use your code, I encountered a problem. Why is there no "utils" in my "helpers"?

**Reply by sheepone12138** [reply]
Hello! I really can't find "bin_packing_model". I hope you can reply when you receive this message. This is relatively important.

**Reply by mmghannam** [reply]
You can find it [here](https://github.com/scipopt/PySCIPOpt/blob/38fb332b8c19ea6af937868f71f223c169405e66/tests/helpers/utils.py#L107), this is just the helper code for our tests. But you could replace it with the model that you wanted to attach the event handler to.

**Reply by Joao-Dionisio** [reply]
Oh, I'm sorry @sheepone12138 I thought I had answered your other question. The bin packing model isn't important, it's supposed to be the model you want to see the log of.

But the helper models are [here](https://github.com/scipopt/PySCIPOpt/tree/master/tests/helpers) in the tests folder.
